### DISCUSSION TITLE
How to mass hide/unhide BIRT controls

### DISCUSSION THREAD

**Discussion by raviprsd18** [eclipse-birt/birt#1568] [Q&A] [answered]
Hello BIRT team,

I have many report design files that contains lot of controls inside each report design.
Following is my problem statement.

1. When ever any validation or any exception occurs in the report execution, I need to show a friend message to the users inside the report. ( Reason for the exception is logged for future troubleshooting )
2. I need a better way to hide all the controls in the report design and show only error message the user ( attached the sample )
![image](https://github.com/eclipse-birt/birt/assets/46317080/badc8f83-5bf9-4f58-9305-172938108512)

I don't want to put all my controls inside a grid and hide/show that grid because this resulting in some formatting issue when I export to .XLSX file type.

Any idea on achieving this ?

Thanks in advance.

**Comment by speckyspooky** [comment]
Can you give an example what you mean with controls in report currently I have no idea
what your use case is and about which controls we are speak.

If you would hae a very small test example whith your topic would be help to see the problem detailed.

**Comment by raviprsd18** [comment]
Hello @speckyspooky thanks for the reply.

Here I am attaching a sample report that contains label, text, image, table BIRT controls ( all are dummy controls, just for explanation purpose ), here controls means BIRT controls like label, dynamic text, image etc.
I have also added a grid control whose name _ErrorMessageTemplate_. 

I need help in showing only _ErrorMessageTemplate_ grid when ever exception happens in the report by hiding all the other controls. In the similar lines if there are no exceptions happens during report execution, I need to hide the _ErrorMessageTemplate_ grid and show rest of the BIRT controls used in report design.
[Example_MassHide_Unhide.zip](https://github.com/eclipse-birt/birt/files/14392154/Example_MassHide_Unhide.zip)

**Reply by speckyspooky** [reply]
Ok, I reviewed your report. I doesn't know exactly which situation should be occure that you are not able to fetch the error before.
Because if a dataset is not executable then you will get an uncatchable exceptionon the execution moment.

If you have a special situation on a script where you will use to catch errors with try-ctach-mechanism then there would be a way.
You can define on report-level at initialization-method of the report a global variable "executionStatusOk" (=true).

If you have a problem then you can set the "executionStatusOk" (= false).

The next step would be to set this variable on the "Visibility"-condition of your ErrorMessageTamplate
and also such check into the "Visibility"-condition of the other statements.

Alternative you can use also a java-script-function which wil be used at "Visibility"-level
if you have further conditions which would set the status to "failed". The way with such function would be better like copy all the stuff directly on the visibility-condition.Ok, I reviewed your report. I doesn't know exactly which situation should be occure that you are not able to fetch the error before.
Because if a dataset is not executable then you will get an uncatchable exceptionon the execution moment.

If you have a special situation on a script where you will use to catch errors with try-ctach-mechanism then there would be a way.
You can define on report-level at initialization-method of the report a global variable "executionStatusOk" (=true).

If you have a problem then you can set the "executionStatusOk" (= false).

The next step would be to set this variable on the "Visibility"-condition of your ErrorMessageTamplate
and also such check into the "Visibility"-condition of the other statements.

Alternative you can use also a java-script-function which wil be used at "Visibility"-level
if you have further conditions which would set the status to "failed". The way with such function would be better like copy all the stuff directly on the visibility-condition.

**Reply by raviprsd18** [reply]
Hello @speckyspooky, 
Thanks for the reply.

I got your point on using the global variable to set it during any error, and use this condition in visibility property of the controls.
This is what I don't want to do, with this design logic, I need to go to all the controls visibility property and write the code based on global variable. 

I am looking some generic solution to mass hide/unhide controls that should eliminate maintenance problem i.e., if new controls are added to the report; again we need to write the logic to the new controls visibility property to hide/unhide based on the global variable.

You were also asking in which scenario you are not able to catch the _error_ before. 
**Let me take an example**: Say I have a report parameter that takes _age_ as a input from the user, but user enters alphabets instead of numbers. Now I need to show the user that he/she have entered wrong input, the expected input for age field is number ( _kind of validation message showing inside the report by hiding all other controls_ )

**Reply by speckyspooky** [reply]
Here you will get a draft if you try to handle parameter topics.
By default for the age normally you would use integer instead of string, but I agree the age is invalid for negativ numbers.

I have added an error simulator which will be set a hidden parameter-variable into the validation of the age-parameter
(only for simulation of your target validation rule).

If the value is "failed" then a script on beforeFactory will check your result "failed" / "ok".
Dependen on the status all elements can be removed/droped without the master-error-template.

I work with the Name of the BIRT-elements to fetch the elements and remove them.
So it will be avoid that each single element (e.g. table content) will be deleted step by step.

### Here is an example
[Example_MassHide_Unhide_TG.zip](https://github.com/eclipse-birt/birt/files/14409645/Example_MassHide_Unhide_TG.zip)

**Reply by raviprsd18** [reply]
Hello @speckyspooky, 
Thanks for the reply.

I was looking for this kind of solution. Last request, in the code I can see all the elements names are stored in the array then based on the reportParameter _success / failure_ controls are dropped.

Do BIRT have any API or inbuilt function that give all the controls reference present in the reportDesign ? So that I can eliminate the logic of storing the control names in array and then dropping it. If this is possible then in future if any new controls are added or deleted no need to update the array, as a result easy to maintain the code.

**Reply by hvbtup** [reply]
That is the DE API.
You can inspect and modify a report design file from within Javascript in the `beforeFactory` event, IIRC.
But honestly, I think the way you are approaching this topic is wrong.
Usually it should be much easier to e.g. check your parameter values in the `beforeOpen` event of the outermost (regarding the bindings) dataset and just "not find" any top-level data if the parameters are invalid, e.g. by setting the dataset parameters to a combination that cannot find anything.

**Reply by hvbtup** [reply]
You can also fill an array variable (let's call it `errorMessages`) along the way, and add a scripted dataset (which can be very generic, e.g. referenced from a library) to the end of the report which shows these error messages.

**Reply by speckyspooky** [reply, marked as answer]
Yes, you can get the IDs instead of a static array based on the design handle.
Here a small example for the beforeFactory-method:

```
var body = reportContext.getDesignHandle().getBody();
var it = body.iterator();

glElemOverview = "";
while (it.hasNext()) {
       var obj = it.next();
       glElemOverview += "<br>" + obj.getName();
}
```

**Comment by raviprsd18** [comment]
@speckyspooky thanks for the solution, I got exactly what I need. As always you saved me :)
